src/Npgsql.Json.NET/Internal/JsonNetJsonConverter.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Npgsql.Internal;
+using JsonSerializer = Newtonsoft.Json.JsonSerializer;
+
+namespace Npgsql.Json.NET.Internal;
+
+sealed class JsonNetJsonConverter<T> : PgStreamingConverter<T?>
+{
+    readonly bool _jsonb;
+    readonly Encoding _textEncoding;
+    readonly JsonSerializerSettings _settings;
+
+    public JsonNetJsonConverter(bool jsonb, Encoding textEncoding, JsonSerializerSettings settings)
+    {
+        _jsonb = jsonb;
+        _textEncoding = textEncoding;
+        _settings = settings;
+    }
+
+    public override T? Read(PgReader reader)
+        => (T?)JsonNetJsonConverter.Read(async: false, _jsonb, reader, typeof(T), _settings, _textEncoding, CancellationToken.None).GetAwaiter().GetResult();
+    public override async ValueTask<T?> ReadAsync(PgReader reader, CancellationToken cancellationToken = default)
+        => (T?)await JsonNetJsonConverter.Read(async: true, _jsonb, reader, typeof(T), _settings, _textEncoding, cancellationToken).ConfigureAwait(false);
+
+    public override Size GetSize(SizeContext context, T? value, ref object? writeState)
+        => JsonNetJsonConverter.GetSize(_jsonb, context, typeof(T), _settings, _textEncoding, value, ref writeState);
+
+    public override void Write(PgWriter writer, T? value)
+        => JsonNetJsonConverter.Write(_jsonb, async: false, writer, CancellationToken.None).GetAwaiter().GetResult();
+
+    public override ValueTask WriteAsync(PgWriter writer, T? value, CancellationToken cancellationToken = default)
+        => JsonNetJsonConverter.Write(_jsonb, async: true, writer, cancellationToken);
+}
+
+// Split out to avoid unneccesary code duplication.
```

Comment:
Yes we're sure :)

Reply:
OK :+1:

Am surprised given how trivial it seems, any idea if it's tracked somewhere?

Reply:
Yeah it was reiterated in Toubs post as well. I agree it's certainly not a very idiomatic coding style...

>[dotnet/runtime#81603](https://github.com/dotnet/runtime/pull/81603) moved some code around to reduce how much code was in a generic context. With Native AOT, with type parameters involving value types, every unique set of type parameters used with these collections results in a unique copy of the code being made, and with all of the various strategies around just in case they’re necessary to optimize a given set, there’s potentially a lot of code that gets duplicated. This change was able to shave ~10Kb off each generic instantiation.

It's not the easiest thing for the JIT to figure out without making the feature catch only trivial cases. One static method not using any generics calling another that might do typeof(T) etc complicates simple single method analysis.

I certainly hope once they settled in with NAOT they'll look at improving these kinds of things though.

Relevant to note is the undocumented `<IlcFoldIdenticalMethodBodies>true</IlcFoldIdenticalMethodBodies>`, which does solve this, but it breaks certain reflection patterns and stack traces afaik.
There has been some discussion about making that feature fully fledged and supported (https://github.com/dotnet/runtime/issues/88748).

Reply:
Thanks for the links and added info!